airflow/dag_processing/manager.py

```diff
@@ -1129,7 +1129,6 @@ def _kill_timed_out_processors(self):
                 )
                 Stats.decr('dag_processing.processes')
                 Stats.incr('dag_processing.processor_timeouts')
```

Comment:
```suggestion
                Stats.incr('dag_processing.processor_timeouts')
                # Deprecated. Remove when we decide to remove deprecations
```

Reply:
I like a message like this, more informational rather than a TODO with a defined time. What about this:
```
Deprecated. If and when another major version of Airflow is released, this should be removed.
See https://github.com/apache/airflow/pull/27067#issuecomment-1289736364
```

Reply:
I think we might arrive at some point that we remove some deprecations even if we are "semVer" compliant :).  We do it for example when we have security fixes and generally good reason for it.

Reply:
Hmm, I'm very curious about that! As a developer that's a bit of a trust-buster to have user facing API break in a non-major release. But that's a philosophical discussion for another time :stuck_out_tongue_closed_eyes: 

Let me take another stab at something more generic:
```
Deprecated. If and when another major version of Airflow is released, or when
deprecated code is being otherwise culled, this should be removed.
See https://github.com/apache/airflow/pull/27067#issuecomment-1289736364
```

Reply:
all suggestions seems a bit confusing.

maybe we need to adopt a policy and say `deprecated and subject to removal per core deprecation policy` or something...

or, short of that.... maybe just

```
Deprecated; will be removed in a future Airflow release.
```

If we don't know _when_ it will be removed, then there's not much value in saying anything about when (i.e. "when we decide to" or "if we ever have another major release".

Or could say

```
Deprecated; may be removed in a future Airflow release.
```

i.e. may instead of will.

But, really that's what deprecated means -- it means it _will_ be removed, so stop using it, no? so let's just call it like it is.

Reply:
Yes. Very much agree with the last proposal from @dstandish . There is no particular need to define in which exact it will be removed. We have no "crystal ball" to tell us the future. Whatever we write there might be a lie. I prefer to be vague rather than potentialy lie. It might well be that at some point in time we relax a bit our approach to backwards compatibiity and we decide that we will allow some, rarely used features to be removed. 

Context:

Backwards compatibility and deprecation removal and SemVer might seem like 0/1 game but it isn't. Not even close. Whether something is classified as backwards incomatibility / major version bump (in terms of SemVer versioning) or not highly depends on many factors but most of all it should be a decision made by release manager (based on consensus of the community/PMC members). The code does not decide for us. For example we might decide to disable some feature (we did in the past) when there is a security vulnerabiliyt. Does it break someone's workkflow? Probably yes. SHOULD it break someone workflow ? Probably YES. Is it backwards compatible? Technically yes. Should we bump major version of Airfow to 3 when we do it ? Definitely NOT. 

For me backwards compatibility is more close to quantum phusics than binary choices. It's all about probabilty that we will break majority of our user's workflows. Every single release of ours (whether major, minor or patchlevel) breaks someone workflow and this is absolutely inevitable.

This is very well captured in Hyrum's Law: https://www.hyrumslaw.com/ and in the obligatory XKCD: 

> With a sufficient number of users of an API, it does not matter what you promise in the contract: all observable behaviors of your system will be depended on by somebody.

![image](https://user-images.githubusercontent.com/595491/200565569-e84bc0c6-a65b-4df9-aaa0-575af2c0b14e.png)

IMHO - what we treat as "backwards incompatible" change should be solely based on our assesment on how big impact it has on our user's migration and whether it is easy to recover for them. It should not be solely based on the actual 'removal" of some feature or changing the defaults. I think this is a very narrow-scoped approach that is not even correct assesment in the first place. You WILL break other's workflows with every release. It's inevitable. Pretending that it's different is just not coping well with reality (which is very far from 0/1 binary choices).

After quite some time of thinking on how we can approach deprecations, I start to think more and more that we should be slightly more liberal when it comes to some deprecations and remove them earlier. The goal of adopting SemVer was not the "versioning" strictness. The main reason we adopted it was because we wanted to make our users comfortable in knowing that their migration will be - very likely - a painless or very easy to handle - experience. We cannot guarantee painless (because of Hyrum's Law). But we can make it super-easy to handle. Goal achieved.

We just need to understand the consequences and deliberately decide, rather than blindly rely on "removed/changed" feature. Blind 0/1 decisions based on code added/removed is goting an easy (and wrong IMHO) path. Nothing frees us from thinking and making decisions. Community over the code is the ASF motto. If we put everything as fully automated rule, we do not need PMC and community any more. Those are guidelines for us, but we should have final say as humans how we interpret those. 

INHO - if we do such removal infrequently and we have an easy (possibly even automated or semi-automated or even super-easy to follow by our users) ways how to deal with those, I think we should be able to remove some of the deprecated code without actually breaking our SamVer promises. Something that we will likely be discussing in the future. And I will advocate for that in the future rather than going to Airflow 3.

Reply:
Great discussion everyone! Two followups:

1. Related to @dstandish feedback:
> If we don't know when it will be removed, then there's not much value in saying anything about when (i.e. "when we decide to" or "if we ever have another major release".

The value I'm trying to provided with my proposals isn't to forecast _when_ it will be removed, rather to provide some helpful context, namely some examples of conditions that would trigger it's removal. Simply saying "Deprecated; may be removed in a future Airflow release." is a tautology (as anything deprecated is subject to removal at some point in a future release) and doesn't provide the reader any further context beyond the first word "Deprecated". But as always, if I'm out voted, I'm more than happy to go with your suggestion :relaxed: 

2. Related to @potiuk feedback:

I'm struggling to find where to land here. I've been part of some discussions where I've suggested deprecating early ("slightly more liberal" as you put it) and I've received push back, and in this thread we're going the other way (at all costs don't deprecate, just keep the behaviour and document it's existence as deprecated). This quantum approach means that no one ever quite knows if and when anything can be removed, so it leads to long discussions like this one. Whereas, if you take SemVer very literally then at least everyone knows, more clearly, how to proceed.

If we're choosing to be more flexible and liberal (which I mostly agree with), then I actually re-propose to remove this metric entirely. It was publicly messaged to be deprecated along with a batch of other metrics, all of which were removed but this one was forgotten; you can really think of this change as a bug-fix of sorts (so no major release needed). If we do the math of how many users will really be affected, versus the code quality improvements of not having to maintain and document this behaviour, I think we could easily land on the removal side. Thoughts?

Reply:
I simply try to think and analyse and adapt my understanding of the world. I used to think "backwards compatibillity is easy" and SemVer provides answers. But this was "past". Now we are in the "current".

I've changed my mind seeing some real-world effect and Hyrum's Law is simply reflecting my current understanding. 
Maybe that's why @o-nikolas some past discussions and push-back of mine migh have mislead you. The push back was based on my (then) strict interpretation of what "backwards incompatibility" is.  But since then - things happen. 

That includes a user who complained tha their deployment started to throw OOM. Since they had not changed anything between the upgrades, the user questioned backwards compatibility of Airlfow. True story. 

For me that was a bit of revelation and true embodiment of Hyrum's Law and the XKCD comic I quoted above. And while I am pretty much still in the campl of using SemVer as communication tool, I am also more on adding more freedom in interpreting of what "backwards compatibility" is. 

This has been mostly summarized in my "lazy consensus" post from today about providers: https://lists.apache.org/thread/nvzdgshwm7s0c2sr4sbokp8kbfjnc039. Not later than 6 months ago when we released 2.2+ providers I was an advocate of "breaking change" and "bump major versions" of all providers. Now - with 6 more months of experience, listening to users and looking at other projects (and thinking hard about it) I changed my opinion: the changes in min-airflow-version are unlikely to break someone's workflow, so let's not treat them as breaking. Minor version upgrade is enough.

The most important learning is that there is no "objective" criterium for breaking/non-breaking. Something that is not breaking for me, might be breaking for others. And there is very litttle we can do about it. Realisation of that is pretty mind-boggling and ot made me change my mind when it comes to "breaking" interpretation.

This issue here is siimilar:

* Is SemVer still valid for us ? I think yes. 
* But - can we remove this metrics without bumping major version? I think yes. 


While those two seem contradicting at first, they are not (in my head). While I think breaking changes should bump major versions, such removal is unlikely to break someone workflow, so it is not breaking.

Bottomline:

Yes I think we should remove this metrics now. That would be my vote now.

Proposal:

Now, how we can do some actionable steps from that statement ?

I think there is a way. We could raise a thread on devlist to see if that interpretation of "breaking change" is something that the community is ok with. I think we've never agreed on "what breaking change is". And while it seems obvious - it is not. What's breaking for me might not be necessarily breaking for you.

We could ask on the devlist whether the community is ok with "relaxed"  interpretation of "breaking change" - not every removal is breaking, not every parameter change is breaking. Even if technically there are some workflows that will be broken. If we assess that a change is not likely to break many user's workflow - we shoudl be able to treat removals like that as non-breaking.

While this seems like a nuance and unnecessary discussion, I think it is an important decision if we can make it. This will mean that some of the old stuff that we are pretty sure will not wreak havoc with our customers, can be removed without going to Airlfow 3 - while still keeping SemVer promises (interpreted SemVer, yes, but still showing that we promise "easy migration" for all Airflow 2 versions). Eventually it might mean that we will never have Airflow 3 (which I think is not going to happen anyway). And that we will gradually get rid of the deprecations in Airflow 2 rather than cumulating them.
 
Another option - if we will not be able to get to consensus - is to get rid of the strict SemVer compliance and propose our own non-fully-semver compatible versioning with some more relaxed meaning of "breaking".

I am willing to start such a thread in the next few days. I see in a way what happens with the 2.3+ providers as a test for it. This is an "interpretation" of the SemVer for providers - very similar to the above line of thoughts. If people will be ok with the "non-breaking" 2.3+ provider change, it's a good sign that we can likely apply more relaxed approach for Airflow core as well.

@o-nikolas, @dstandish - is that clearer now where I am coming from ? Is that something that we can agree on as a direction :) ?

Reply:
@potiuk 
> is that clearer now where I am coming from ? Is that something that we can agree on as a direction :) ?

Yupp, and creating a dev email list discussion seems like a reasonable next step to me. Perhaps the approach we decide on should even be enshrined in an AIP (even though that might be unconventional).

Ultimately, the most important things to me are to 1) make a group/community decision on the direction (whether that's strict SemVer, our own approach, something in between) 2) document that direction as explicitly and detailed as possible, then 3)  stay consistent with whatever we decide.
2 and 3 combined will reduce developer overhead and long discussions each time this question comes up (since it will always come up a lot).